akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala

```diff
@@ -473,31 +481,38 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
     private var outstandingStreamWindow: Int = Http2Protocol.InitialWindowSize // adapt if we negotiate greater sizes by settings
     outlet.setHandler(this)
 
-    def onPull(): Unit = dispatchNextChunk()
+    def onPull(): Unit = incomingStreamPulled(streamId)
     override def onDownstreamFinish(): Unit = {
       debug(s"Incoming side of stream [$streamId]: cancelling because downstream finished")
       multiplexer.pushControlFrame(RstStreamFrame(streamId, ErrorCode.CANCEL))
       // FIXME: go through state machine and don't manipulate vars directly here
       streamStates -= streamId
+      wasClosed = true
+      buffer = ByteString.empty
+      trailingHeaders = None
     }
 
-    def onDataFrame(data: DataFrame): Option[StreamState] = {
-      if (data.endStream) wasClosed = true
+    def isDone: Boolean = outlet.isClosed
 
-      outstandingStreamWindow -= data.sizeInWindow
-      if (outstandingStreamWindow < 0) {
+    def onDataFrame(data: DataFrame): Unit =
+      if (wasClosed) {
         shutdown()
-        multiplexer.pushControlFrame(RstStreamFrame(streamId, ErrorCode.FLOW_CONTROL_ERROR))
-        // also close response delivery if that has already started
-        multiplexer.closeStream(streamId)
-        Some(Closed)
```

Comment:
so the transition to `Closed` is no longer signaled from here.

How does it go into that state now? Is it an effect of `shutdown` failing the outlet?

Reply:
It's now using `afterBufferEvent` consistently to figure out the next state after each call into `IncomingStreamBuffer` where before this return value was sometimes used and sometimes ignored.

Reply:
right, which looks at `buffer.isDone` which checks `outlet.isClosed` :+1: